src/libsyntax/parse/parser.rs

```diff
@@ -4070,6 +4088,9 @@ impl<'a> Parser<'a> {
                     TraitBoundModifier::None
                 };
                 bounds.push(TraitTyParamBound(poly_trait, modifier));
+                if has_parens {
+                    self.expect(&token::CloseDelim(token::Paren))?;
+                }
             } else {
                 break
```

Comment:
@petrochenkov can you add some comments about this, at minimum? moving the `expect(closing_patern)` also sounds good.